src/fides/api/util/tcf_util.py

```diff
@@ -0,0 +1,637 @@
+) -> None:
+    """Either appends or replaces the legal_bases, depending on whether replace=True or False
+
+    purpose record updated in place
+    """
+    if not isinstance(purpose_record, TCFPurposeRecord):
+        return
+
+    legal_bases: List[str] = (
+        [legal_basis_for_processing] if legal_basis_for_processing else []
+    )
+
+    if replace:
+        purpose_record.legal_bases = legal_bases
+    else:
+        if legal_basis_for_processing not in purpose_record.legal_bases:
+            purpose_record.legal_bases.extend(legal_bases)
+
+    purpose_record.legal_bases.sort()
+
+
+@lru_cache()
+def get_tcf_contents(
+    db: Session,
+) -> TCFExperienceContents:
+    """
+    Returns the base contents of the TCF overlay.
+
+    Queries for systems/privacy declarations that have a relevant GVL data use and a legal basis of Consent or Legitimate interests,
+    and builds the TCF Overlay from these systems and privacy declarations.
+
+    TCF Overlay has Purpose, Special Purpose, Feature, and Special Feature Sections, that have relevant systems and vendors embedded underneath.
+    The reverse representation is also returned, and Vendors and Systems each have relevant Purpose, Special Purpose, Feature, and Special Feature sections.
+    """
+    system_map: Dict[str, TCFVendorRecord] = {}
+
+    matching_privacy_declarations: Query = get_matching_privacy_declarations(db)
+
+    # Collect purposes and update system map
+    (
+        purpose_map,
+        updated_system_map,
+    ) = build_purpose_or_feature_section_and_update_system_map(
+        PURPOSE_DATA_USES,
+        tcf_component_name="purposes",
+        system_map=system_map,
+        matching_privacy_declaration_query=matching_privacy_declarations,
+    )
+
+    # Collect special purposes and update system map
+    (
+        special_purpose_map,
+        updated_system_map,
+    ) = build_purpose_or_feature_section_and_update_system_map(
+        SPECIAL_PURPOSE_DATA_USES,
+        tcf_component_name="special_purposes",
+        system_map=updated_system_map,
+        matching_privacy_declaration_query=matching_privacy_declarations,
+    )
+
+    # Collect features and update system map
+    (
+        feature_map,
+        updated_system_map,
+    ) = build_purpose_or_feature_section_and_update_system_map(
+        [feature.name for feature in GVL_FEATURES.values()],
+        tcf_component_name="features",
+        system_map=updated_system_map,
+        matching_privacy_declaration_query=matching_privacy_declarations,
+    )
+
+    # Collect special features and update system map
+    (
+        special_feature_map,
+        updated_system_map,
+    ) = build_purpose_or_feature_section_and_update_system_map(
+        [feature.name for feature in GVL_SPECIAL_FEATURES.values()],
+        tcf_component_name="special_features",
+        system_map=updated_system_map,
+        matching_privacy_declaration_query=matching_privacy_declarations,
+    )
```

Comment:
Sure, this could be moved to fideslang in some form eventually.  One thing I want to highlight that is happening here though is that each Feature and Vendor, for example, are not being created independently.  We might create two Features and update the corresponding Vendor to nest those features beneath the vendor.  Likewise the individual Features have a vendors section where the vendor id is added.  The TCF components affect each other's construction.  

Not to say there isn't value in a Privacy Declaration -> Purpose/Feature/Vendor mapping, just that those constructs can't live in isolation. Other PrivacyDeclarations will also update the constructs created by a previous PrivacyDeclaration.

Reply:
I agree with you happy to help figure out how to consolidate more of this in fideslang as a next step!